sum.go

```diff
@@ -1,241 +1,49 @@
 package multihash
 
 import (
-	"crypto/md5"
-	"crypto/sha1"
-	"crypto/sha512"
 	"errors"
 	"fmt"
-
-	blake2b "github.com/minio/blake2b-simd"
-	sha256 "github.com/minio/sha256-simd"
-	murmur3 "github.com/spaolacci/murmur3"
-	blake2s "golang.org/x/crypto/blake2s"
-	sha3 "golang.org/x/crypto/sha3"
 )
 
 // ErrSumNotSupported is returned when the Sum function code is not implemented
 var ErrSumNotSupported = errors.New("Function not implemented. Complain to lib maintainer.")
 
 var ErrLenTooLarge = errors.New("requested length was too large for digest")
 
-// HashFunc is a hash function that hashes data into digest.
-//
-// The length is the size the digest will be truncated to. While the hash
-// function isn't responsible for truncating the digest, it may want to error if
-// the length is invalid for the hash function (e.g., truncation would make the
-// hash useless).
-type HashFunc func(data []byte, length int) (digest []byte, err error)
-
-// funcTable maps multicodec values to hash functions.
-var funcTable = make(map[uint64]HashFunc)
-
 // Sum obtains the cryptographic sum of a given buffer. The length parameter
 // indicates the length of the resulting digest and passing a negative value
 // use default length values for the selected hash function.
 func Sum(data []byte, code uint64, length int) (Multihash, error) {
-	if !ValidCode(code) {
-		return nil, fmt.Errorf("invalid multihash code %d", code)
+	// Get the algorithm.
+	hasher, err := GetHasher(code)
+	if err != nil {
+		return nil, err
 	}
 
+	// Feed data in.
+	hasher.Write(data)
+
+	// Compute hash.
+	//  Use a fixed size array here: should keep things on the stack.
+	var space [64]byte
+	sum := hasher.Sum(space[0:0])
+
+	// Deal with any truncation.
 	if length < 0 {
 		var ok bool
 		length, ok = DefaultLengths[code]
```

Comment:
Yeah, I just didn't touch this yet.  Input wanted.

Option 3 seems bad to me, Options 1 / 2 / 4 all seem potentially viable.

Reply:
Maaaybe related to https://github.com/multiformats/go-multihash/issues/137 (if we choose Option 4).

Reply:
As pointed out in https://github.com/multiformats/go-multihash/issues/137#issuecomment-791516068 at least the case of DefaultLengths seems easy since we can use `hash.Size()`. That probably means the DefaultLengths map can be dropped as it is replaceable with `h, err: = GetHasher(code) ... err things ... h.Size()`

Reply:
Right, yeah, using `Hash.Size` does seem to do the thing here.

I drafted the diff that removes `DefaultLengths` and it's very small and clean, so I suggest we go with dropping it.  It seems very unlikely anyone else will have been touching that map directly.

Reply:
... I take it back, there's one more usage I didn't see:

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/opts/opts.go#L113-L115

- (Sidenote) That's an ...interesting... piece of code
  - Not sure quietly setting the value to a smaller number is what I'd expect from that.  (But not necessarily here to argue about it either.)
  - Not sure "default length" and "max length" are really the same concepts.  (Meh, naming.)
- It's wanting to get this size info but not actually about to do any hashing yet, so it would rather not experience an allocation just to get the property information.

Thoughts?  Is this worth having an length info map for, or should we do something else here?